README.md
# Raiden Service Bundle (RSB)

## What is this repository

This repository contains the documentation and configuration necessary to run a
Raiden Service Bundle.

**Current release:** [Latest](https://github.com/raiden-network/raiden-service-bundle/releases/latest)

## Table of Contents

- [Overview](#overview)
- [Requirements](#requirements)
  - [Hardware](#hardware)
  - [Software](#software)
- [Installation](#installation)
- [Upgrades](#upgrades)
- [Known issues](#known-issues)
- [Contact / Troubleshooting](#contact-troubleshooting)
- [Changelog](#changelog)

## Overview

The Raiden Network uses a federation of [Matrix](https://matrix.org) servers as
its transport layer and a set of services for improved usability. This set of
services is called the Raiden Service Bundle. To ensure reliability,
availability and neutrality it is desirable that those services are being
operated by multiple independent entities.

Therefore we provide this repository which allows easy setup of such a service
bundle. It uses docker and docker-compose for easy installation and upgrades.

Currently only this single-server configuration is supported; in the future we
may also provide configurations with services split among multiple servers.

### Used software

- docker
- docker-compose
- Synapse
- Postgres
- Traefik
- Raiden Services (Pathfinding, Monitoring)

### Structure


```
+-------------------+
|                   |
|   Raiden clients  |
|                   |
+---------+---------+
          |https://
==========|==========
          |
+---------v---------+                       Federation to
|                 +-+-------------------->  other Raiden
|      Traefik    | |                       Matrix servers
|                 |-+---------+----------------------+
+---------+-------+-+         |                      |
          |       |           |                      |
+---------v-------v-+   +-----v----------------+ +---v-----------------+
|                   |   |                      | |                     |
|      Synapse      |   |  Raiden Pathfinding  | |  Raiden Monitoring  |
|                   |   |                      | |                     |
+---------+---------+   +-------------------+--+ +-+-------------------+
          |                                 |      |
+---------v---------+                     +-v- - - v -+
|                   |                     |
|     Postgres      |                        ETH_RPC  |
|                   |                     |
+-------------------+                     + - - - - - +
```


We use Traefik as a reverse proxy and also utilize its capability of
automatically provisioning Let's Encrypt TLS certificates.

The Synapse server is being run in the so-called split worker configuration
which increases throughput.

The database stores the message data. Since the transport layer is considered
ephemeral in Raiden it is not necessary to arrange for backups of the database
data.

### Network

After a successful deployment the following ports will be in use:

- 80 - HTTP
  - Redirects to HTTPS
  - Let's Encrypt HTTP challenge for certificate provisioning
- 443 - HTTPS
  - Synapse (on subdomain `transport.$<SERVER_NAME>`)
    - Client API access
    - Server-to-Server federation
  - Raiden Pathfinding Server (on subdomain `pfs.$<SERVER_NAME>`)
  - Metrics export (IP restricted, see below)

## Requirements

### Hardware

Minimum recommended for a production setup:

- 16 GiB RAM
-  8 Cores
- 50 GiB SSD

Note: The default Postgres configuration assumes 16GiB of system RAM

### Software

- Docker >= 17.12
- docker-compose >= 1.21.0

### Other

- A domain (or subdomain) for exclusive use by this server
- To ensure acceptable performance the server should be reserved for exclusive use by the RSB.

## Installation

> **Note**
>
> Running an RSB provides important infrastructure to users of the Raiden
> Network and failures potentially affect a lot of users. It is therefore
> recommended to have devops experience and basic knowledge of the Ethereum
> ecosystem in order to run an RSB. Also a minimum commitment of a reasonable
> response time in case of outages is required.

### Preparation

1. Provision a server that meets the [hardware](#hardware) and [software](#software) requirements listed above.
1. Ensure a domain (or subdomain) is available

   Examples:
   - raiden.somedomain.com
   - raiden-service-bundle-somecompany.tld

1. Configure `A` (and optionally `AAAA`) DNS records for the domain pointing to the servers IP address(es)
1. Configure a `CNAME` DNS record for `*.<domain>` pointing back to `<domain>`


**NOTE:**

If you intend to use a subdomain it is important to be aware of the security implications.
Subdomains share Cookies and Browser LocalStorage with the apex domain.
Therefore we strongly suggest that a subdomain is only used below an apex domain that does *not*
host an application that relies on either Cookies or LocalStorage for security relevant purposes (e.g. user authentication).


### Installing the RSB

**NOTE:**
This document will sometimes display release candidate versions, also known as pre-releases in the section below. You
can identify this, if there is an `rcX` at the end of the version (E.g. `2019.03.0rc5`). When in doubt, always check against the
[latest full release](https://github.com/raiden-network/raiden-service-bundle/releases/latest). If the version is
different from what you see below, you should stick to the "full release" and replace the version accordingly.

1. Clone the [current release version of this repository](https://github.com/raiden-network/raiden-service-bundle/tree/2022.05.0)
   to a suitable location on the server:

   ```shell
   git clone -b 2022.05.0 https://github.com/raiden-network/raiden-service-bundle.git
   ```
1. Copy `.env.template` to `.env` and modify the values to fit your setup. Please read [Configuring the `.env` file](#configuring-the-env-file) for detailed information.
   - We would appreciate it if you allow us access to the monitoring interfaces
     (to do that uncomment the default values of the `CIDR_ALLOW_METRICS` and `CIDR_ALLOW_PROXY` settings).
   - We also recommend that you provide your own monitoring. The setup of which is currently out of scope of this document.
   - Please read the disclaimers for the path finding and monitoring services carefully and uncomment the variables `<SERVICE>_ACCEPT_DISCLAIMER` if you agree. Note, that without agreement the services won't start.
1. If you haven't done so before, run `./register-service-provider.sh register` (it uses configuration values from `.env`). Please read the information provided [Registering as a RSB Provider](#registering-as-a-rsb-provider) carefully before executing the script.
1. Run `docker-compose up -d` to start all services
   - The services are configured to automatically restart in case of a crash or reboot

**NOTE:**

After a new RSB has been registered and added to the
`known_servers/known_servers-production-v3.0.0.json` file it can take up to 24
hours for the information to propagate to existing RSB installations.

During this time some services will not yet be able to start successfully and
log various error messages. This is expected behaviour and will resolve itself.

After the 24h have elapsed all services should run successfully.
See [verifying that the RSB is working](#verifying-that-the-rsb-is-working) below.

### Configuring the `.env` file
After cloning the repository the `.env` file needs to be configured. A template named `.env.template` is provided. Below you find a detailed list of the parameters to be set and their explanations.

- `SERVER_NAME`: The host domain without protocol prefix `https://` respectively
- `LETSENCRYPT_EMAIL`: Email address to use when requesting LetsEncrypt certificates
- `CIDR_ALLOW_METRICS`: Metrics whitelist. IP address/network whitelists for access to non-public parts of the service. Uses CIDR notation. Separate multiple entries with commas. Example values: 10.0.0.0/16,10.1.2.3/32 or 10.1.2.3/32.
- `CIDR_ALLOW_PROXY`: Proxy metrics / management interface whitelist
- `WORKER_COUNT`: Number of worker processes to start, setting this to the number of CPUs is a good starting point
- `DATA_DIR`: Data dir location. Optional, defaults to ./data in the checkout directory
- `URL_KNOWN_FEDERATION_SERVERS`: URL to use to fetch federation whitelist - used only for testing
- `KEYSTORE_FILE`: The keystore file which has to be located in ${DATA_DIR}/keystore
- `PASSWORD`: Password to decrypt the keystore file
- `ETH_RPC`: Ethereum RPC URL. This is used to communicate with the Ethereum blockchain. You can either [host a node yourself](https://ethereum.org/en/developers/docs/nodes-and-clients/run-a-node/) or use a service like [Infura](https://infura.io/). Regardless, the RSB requires a minimum of 200,000 requests per day and potentially more depending on traffic. Please make sure that the chosen service is able to handle this.
- `PFS_ACCEPT_DISCLAIMER`: TRUE or FALSE if you accept the Pathfinding Service disclaimer or not. Read the Disclaimer [here](https://github.com/raiden-network/raiden-services/blob/f4bcb9c289e093754204fe18684e4b57558ea29b/src/pathfinding_service/constants.py#L34)
- `MS_ACCEPT_DISCLAIMER`: TRUE or FALSE if you accept the Monitoring Service disclaimer or not. Read the Disclaimer [here](https://github.com/raiden-network/raiden-services/blob/f4bcb9c289e093754204fe18684e4b57558ea29b/src/monitoring_service/constants.py#L22)
- `CHAIN_ID`: Chain ID of the connected Ethereum node.
- `PFS_SERVICE_FEE`: The Pathfinding Service Fee to be paid for requests
- `PFS_OPERATOR`: Official Operator Name
- `PFS_INFO_MESSAGE`: Info message. Will be displayed on info endpoint.
- `LOG_LEVEL`: 'INFO' or 'DEBUG' recommended
- `SERVICE_REGISTRY`: The address of the ServiceRegistry contract to use. When running on mainnet, use the address of latest deployed contract which can be found here: https://github.com/raiden-network/raiden-contracts/blob/master/raiden_contracts/data_0.50.0/deployment_services_arbitrum-one.json#L6

### Registering as a RSB Provider
For your newly deployed Raiden Service Bundle to be used by Raiden nodes it must be registered.

1. **Registering in the Services Registry On-Chain**
  - In order to register as a service provider you need to run the script [`register-service-provider.sh`](https://github.com/raiden-network/raiden-service-bundle/blob/master/register-service-provider.sh)` register`.
  - In case you abort the script before finishing the registration process it is possible to continue the process by running the script again.
  - Make sure that you have configured a keystore file (`$KEYSTORE_FILE` in `.env`). If not, the script will exit with an error and you cannot register as a service provider.
  - Make sure you have ETH on the above account. An estimate will be shown during the registration process, before any transactions are sent.
  - Make sure that the configured account has enough RDN funding to register as a service provider.
    You can check the [registry contract](https://etherscan.io/address/0x1F54A809480E5E6349F09aaCaD14ec94C92EecD9#readContract) for the current price of a slot.
    You will find the price under `3. currentPrice`. To get the price in RDN divide the value by (10^18).
    The script will inform you about the price during the registration process as well.
  - If you do not have sufficient RDN, you can either use an exchange of your choice or an on-chain decentralized exchange like [Uniswap](https://app.uniswap.org/#/swap?outputCurrency=0x255aa6df07540cb5d3d297f0d0d4d84cb52bc8e6) to acquire it.


2. **Extending `known_servers/known_servers-production-v3.0.0.json`**
  - In order to be whitelisted in the Matrix Federation, the list needs to be extended with your server name.
  - [Create an issue](https://github.com/raiden-network/raiden-service-bundle/issues/new) and submit the
   domain / URL of the newly deployed server for inclusion in the list of known servers.
   Please, state your server name as you have set `$SERVER_NAME` in your `.env` file.
  - It may take up to 24 hours for the federation to accept the server as a new member. Please note, that until this moment, the pathfinding service and monitoring service cannot run properly as they need to use the broadcasting rooms. Once the new server is accepted as part of the federation, all services will restart automatically.


#### Interacting with the service registry contract
Besides the subcommand `register` which can be used to register as a RSB provider, there are several other subcommands to interact with the service registry contract.

Commands:

- *extend*    Extend the duration of a service registration
- *info*      Show information about current registration and deposits
- *register*  Registers the address of a service deployment with the `ServiceRegistry`.
- *withdraw*  Withdraw tokens deposited to the ServiceRegistry.

You can call [`register-service-provider.sh`](https://github.com/raiden-network/raiden-service-bundle/blob/master/register-service-provider.sh)` <command>` to use them.

### Verifying that the RSB is working

Check the status of the services by executing `docker-compose ps`.
If any services are in a state other than `Up`, `Up (healthy)` or `Exit 0` after the elapse of the 24h waiting period a configuration problem is the most likely cause.
See [troubleshooting the RSB installation](#troubleshooting-the-rsb-installation) below in that case.

- Matrix
  - Check that the following endpoints return a successful response (HTTP status 200):
    - `https://transport.<SERVER_NAME>/_matrix/client/versions`

- PFS
  - Check that the `latest_committed_block` is increasing regularly:

    `docker-compose logs --tail 100 pfs | grep latest_committed_block`
  - Check that the following endpoint returns a successful response (HTTP status 200):
    - `https://pfs.<SERVER_NAME>/api/v1/info`

- MS
  - Check that the `latest_confirmed_block` is increasing regularly:

    `docker-compose logs --tail 100 ms | grep latest_confirmed_block`


### Troubleshooting the RSB installation

If you experience any unexpected behavior while installing the RSB, please do not hesitate to contact the development team. The fastest way to reach out to us is via the public [Raiden Gitter channel](https://gitter.im/raiden-network/raiden).
Otherwise, you can also open an issue in this repository with the predefined template for a [bug report](https://github.com/raiden-network/raiden-service-bundle/issues/new?template=bug_report.md)


## Upgrades

To upgrade to a new release please refer to the [`upgrading document`](./UPGRADING.md) for any
necessary configuration changes.

Afterwards run the following commands:

```shell
git fetch origin --tags
git reset --hard <new-release-tag>
docker-compose pull
docker-compose up -d --remove-orphans
```
## Notes:

 - A 'purger' service will run once a day, removing inactive users from global rooms
  to save disk space and processing performance.
 - If necessary it will restart the ``synapse`` service to fetch an up-to-date
  whitelist of servers.

## Known issues

### Protection against Spam / (D)DoS attacks

There is currently only some protection against Spam and / or DDoS attacks.
This will be addressed in future updates.

### Known servers

The known servers the Raiden clients try to connect to are currently tracked in
the *.yml files in this repository. These lists are used by Raiden clients when

CHANGELOG.md
## Changelog
- 2022-05-02 - `2022.05.0` - **Upgrade Release**
  - Updated `raiden-services` to `v2.0.0`
  - Updated `raiden-synapse-modules` to `0.1.4`
  - Removed `raiden` dependency for synapse image
  - Fix synapse configuration file format
- 2021-12-06 - `2021.12.0` - **Release Candidate**
  - Upgrade to Raiden 3.0.0rc8
- 2021-07-01 - `2021.07.0rc0` - **Release Candidate**
  - Upgrade to Synapse 1.37.1
  - Move more modules to `raiden-synapse-modules`
- 2021-05-12 - `2021.05.0-rc3` - **Release Candidate**
  - internal module to update federation whitelist
- 2021-05-10 - `2021.05.0-rc1` - **Release Candidate**
  - Upgrade Synapse to 1.33.0
  - Use Presence Router Module to distribute presences to PFS
  - Remove global rooms and Room Ensurer
  - Temporarily switch back to one main process in synapse
  - Allow using subcommands in Service Registry script
- 2021-02-25 - `2021.02.0-rc2` - **Release Candidate**
  - Fix an issue with option handling for `raiden-services`
  - Upgrade Synapse to `1.27.0`
  - Re-enabled retention policies
  - Added more prometheus metrics
  - Removed `state_groups_cleaner`
  - Added `rust-synapse-compress-state` tool
  - Updated `raiden-services` to `v0.14.1`
- 2020-11-25 - `2020.11.2` - **Bugfix Release**
  - **Please check UPGRADING.md**
  - Really upgrade to Synapse `v1.22.1`.
  - Upgrade `raiden-services` to `v0.13.2`
- 2020-11-24 - `2020.11.1` - **Maintenance Release**
  - **Please check UPGRADING.md**
  - ~~Upgrade to Synapse `v1.22.1`~~
  - Add configuration for new federation metrics
  - Increase rate limits
  - Disable `synapse` `retention_policy`
  - Introduce `state_groups_cleaner`
  - Introduce `prometheus` `node_exporter`
  - Increase DB `max_wal_size`
- 2020-10-08 - `2020.10.3` - **Bugfix release**
  - Fix an issue with the release process
- 2020-10-07 - `2020.10.2` - **Bugfix release**
  - Fix a few small bugs
    - Stop Room Ensurer crashing when other servers are unreachable
    - Fix restart policy of DB & Redis metrics containers
- 2020-10-06 - `2020.10.1` - **Bugfix release**
  - Upgrade to Raiden Services `v0.13.1`
    - This fixes a bug with the services metrics
- 2020-10-05 - `2020.10.0` - **Upgrade release**
  - Upgrade Synapse to `v1.19.1`
  - Upgrade Raiden Services to `v0.13.0`
  - Expose additional metrics:
    - Postgres
    - Redis
    - PFS
    - MS
- 2020-06-23 - `2020.06.0` - **Maintenance release**
  - Add upgrade instructions
- 2020-06-22 - `2020.06.0rc1` - **Maintenance release**
  - Fix a bug in the purger code
- 2020-06-18 - `2020.06.0rc0` - **Maintenance release**
  - Update Raiden to 1.1.0
  - Update Raiden Services to 0.11.0
  - Update metrics IP
- 2020-05-20 - `2020.05.0` - **Maintenance release**
  - Update Raiden to 0.200.0rc9
  - Update Raiden Services to 0.10.0
- 2020-05-14 - `2020.05.0rc1` - **Maintenance release**
  - Update Raiden Services to 0.9.0
- 2020-04-15 - `2020.04.0` - **Upgrade release**
  - Documentation has been significantly expanded
  - The Synapse containers are now located on the subdomain `transport.<SERVER_NAME>`
  - Upgrade Traefik to v2.2
  - Upgrade Synapse to v1.10.1
  - Pin the version numbers of required docker / PyPI packages
  - Replace room history purging with Synapse retention policies
  - Add a utility to remove inactive users from global rooms
  - Server side broadcast room creation uses the list of known servers as-is.
    Thereby making the list order stable when new providers are added.
  - Per-Synapse-server admin users are now automatically created
  - Various other improvements
- 2019-11-20 - `2019.11.1` - **Upgrade release**
  - This release fixes some issues with the release process of the previous version `2019.11.0`.
- 2019-11-20 - `2019.11.0` - **Upgrade release**
  - Upgrade Synapse to v1.5.1
  - Use `stable` release from https://github.com/raiden-network/raiden-services
  - Use version tagged public images instead of building locally.
  - Make all bundled software versions easier to maintain (`BUILD_VERSIONS` & `docker-compose.yml::x-versions`).
  - Removed auto registration, added interactive registration script (`./register-service-provider.sh`).
- 2019-10-07 - `2019.10.1` - **Upgrade release**
  - Upgrade https://github.com/raiden-network/raiden-services image to `v0.4.0`
- 2019-10-02 - `2019.10.0` - **Upgrade release**
  - Add https://github.com/raiden-network/raiden-services services to the bundle
  - Upgrade Synapse to v1.3.1
  - Tune Postgres default parameters
  - Merge federation access under regular HTTPS port (443)
    - Port 8448 is no longer needed
- 2018-12-19 - `2018.12.0` - **Maintenance release**
  - purger.py restart improvements
- 2018-10-19 - `2018.10.0` - **Maintenence release**
  - Add new servers to known list
  - Upgrade Synapse to 0.33.7
  - Automatically purge historic state and restart service once a day, removing the need for an external cron service
  - Updated minimum docker and docker-compose version requirements
- 2018-08-09 - `2018.8.1` - **Initial public release**
- 2018-08-02 - `2018.8.0` - **Initial version**


purge/purge.py
import os
import click
import json
import requests
import psycopg2
import datetime
import time
import docker
import yaml
from eth_keyfile import extract_key_from_keyfile, create_keyfile_json
from eth_utils import keccak, encode_hex
from eth_keys import keys
from operator import itemgetter
from urllib.parse import urlparse, urljoin, quote


URL_KNOWN_FEDERATION_SERVERS_DEFAULT = (
    'https://raw.githubusercontent.com/raiden-network/raiden-transport/master/known_servers.yaml'
)
SYNAPSE_CONFIG_PATH = '/config/synapse.yaml'


def eth_sign_hash(data: bytes) -> bytes:
    """ eth_sign/recover compatible hasher

    Prefixes data with '\x19Ethereum Signed Message:\n<len(data)>' """
    prefix = b'\x19Ethereum Signed Message:\n'
    if not data.startswith(prefix):
        data = prefix + b'%d%s' % (len(data), data)
    return keccak(data)


@click.command()
@click.argument('db_uri', envvar='MATRIX_DB_URI')
@click.argument('server', envvar='MATRIX_SERVER')
@click.option(
    '--admin-access-token-file',
    default='./admin_access.json',
    show_default=True,
    type=click.Path(dir_okay=False, writable=True),
    help=(
        'JSON file to retrieve and persist "admin_user" and "admin_access_token". '
        'If present, any other --admin-* option is ignored and these credentials are used.'
    ),
)
@click.option(
    '--admin-user',
    help='Full admin user_id. e.g.: @username:server_hostname.com',
)
@click.option(
    '--admin-password',
    help='Admin user password',
)
@click.option(
    '--admin-private-key',
    type=click.Path(dir_okay=False),
    default='admin_key.json',
    help='Admin ETH JSON private key file to derive admin user and password from.',
)
@click.option(
    '--admin-private-key-password',
    help='Admin ETH JSON private key file password. Prompted if needed and not provided.',
)
@click.option(
    '--admin-private-key-generate',
    is_flag=True,
    flag_value=True,
    help=(
        'Generates a private keyfile, encrypted with password, if one does not exist. '
        'This key shouldn\'t be used for transactions, and is only meant to '
        'ease generating a valid ethereum/matrix admin user.'
    ),
)
@click.option(
    '--admin-private-key-print-only',
    is_flag=True,
    flag_value=True,
    help='Print matrix user and password derived from private key and exit',
)
@click.option(
    '--server-name',
    help=(
        'Custom Matrix server name to be used to calculate user_id from private key. '
        'Defaults to server\'s hostname'
    ),
    envvar='SERVER_NAME',
)
@click.option(
    '--admin-set/--no-admin-set',
    default=True,
    show_default=True,
    help=(
        'If true, sets the user as admin in DB. '
        'Requires --admin-user or --admin-private-key'
    ),
)
@click.option(
    '--keep-newer',
    type=int,
    help='Keep events newer than this number of days',
)
@click.option(
    '--keep-min-msgs',
    type=int,
    help='Keep at least this number of message events per room, regardless of --keep-newer',
)
@click.option(
    '--parallel-purges',
    default=10,
    show_default=True,
    type=int,
    help='Max number of purges to run in parallel.',
)
@click.option(
    '--post-sql',
    type=click.File(),
    help=(
        'Pass a SQL script file as parameter to run it on DB after purging. '
        'Useful to run cleanup scripts, like "synapse_janitor.sql".'
    ),
)
@click.option(
    '--docker-restart-label',
    help='If set, search all containers with given label and, if they\'re running, restart them',
)
def purge(
        db_uri,
        server,
        admin_access_token_file,
        admin_user,
        admin_password,
        admin_private_key,
        admin_private_key_password,
        admin_private_key_generate,
        admin_private_key_print_only,
        server_name,
        admin_set,
        keep_newer,
        keep_min_msgs,
        parallel_purges,
        post_sql,
        docker_restart_label,
):
    """ Purge historic data from rooms in a synapse server

    DB_URI: DB connection string: postgres://user:password@netloc:port/dbname
    SERVER: matrix synapse server url, e.g.: http://hostname

    All option can be passed through uppercase environment variables prefixed with 'MATRIX_'
    e.g.: export MATRIX_KEEP_MIN_MSGS=100
    """
    try:
        session = requests.Session()

        # no --server-name, defaults to server hostname
        if not server_name:
            server_name = urlparse(server).hostname

        # admin_access_token_file has priority over everything else, to avoid re-logins
        if os.path.isfile(admin_access_token_file):
            with open(admin_access_token_file, 'r') as fo:
                admin_user, admin_access_token = itemgetter(
                    'admin_user',
                    'admin_access_token',
                )(json.load(fo))
        else:
            if (
                    not admin_user and
                    not os.path.isfile(admin_private_key) and
                    admin_private_key_generate
            ):
                if admin_private_key_password is None:
                    admin_private_key_password = click.prompt(
                        'JSON keyfile password',
                        default='',
                        hide_input=True,
                    )
                with open(admin_private_key, 'w') as fo:
                    json.dump(create_keyfile_json(
                        os.urandom(32),
                        admin_private_key_password.encode(),
                    ), fo, indent=2)

            # derive admin_user and admin_password from private key
            if not admin_user and os.path.isfile(admin_private_key):
                if admin_private_key_password is None:
                    admin_private_key_password = click.prompt(
                        'JSON keyfile password',
                        default='',
                        hide_input=True,
                    )
                pk_bin = extract_key_from_keyfile(
                    admin_private_key,
                    admin_private_key_password.encode(),
                )
                pk = keys.PrivateKey(pk_bin)

                # username is 0x-prefixed lowercase eth address
                admin_user = f'@{pk.public_key.to_address()}:{server_name}'
                # password is server_name signed with key, with eth_sign prefixed hash
                admin_password_bin = pk.sign_msg_hash(
                    eth_sign_hash(server_name.encode()),
                ).to_bytes()
                admin_password = encode_hex(
                    admin_password_bin[:-1] +
                    bytes([admin_password_bin[-1] + 27])  # v += 27
                )
                if admin_private_key_print_only:
                    click.secho(f'PK to Matrix User:     {admin_user}')
                    click.secho(f'PK to Matrix Password: {admin_password}')
                    return

            if admin_user:
                if not admin_password:
                    admin_password = click.prompt(
                        'Admin user password',
                        default='',
                        hide_input=True,
                    )
                response = session.post(
                    urljoin(server, '/_matrix/client/r0/login'),
                    json={
                        'type': 'm.login.password',
                        'user': admin_user,
                        'password': admin_password,
                    },
                )
                assert response.status_code == 200, f'{response!r} => {response.text!r}'
                admin_access_token = response.json()['access_token']
                with open(admin_access_token_file, 'w') as fo:
                    json.dump({
                        'admin_user': admin_user,
                        'admin_access_token': admin_access_token,
                    }, fo, indent=2)

            else:
                raise RuntimeError('No admin_user nor previous access token found')

        if admin_private_key_print_only:
            # only hit if --admin-private-key-print-only passed, but an access token or admin user
            # was already present, or no --admin-private-key file nor --admin-private-key-generate
            return

        with psycopg2.connect(db_uri) as db, db.cursor() as cur:

            # set user as admin in database if needed
            cur.execute(
                'SELECT admin FROM users WHERE name = %s ;',
                (admin_user,),
            )
            if not cur.rowcount:
                raise RuntimeError(f'User {admin_user!r} not found')
            is_admin, = cur.fetchone()
            if admin_set and not is_admin:
                cur.execute(
                    'UPDATE users SET admin=1 WHERE name = %s ;',
                    (admin_user,),
                )
                db.commit()
            elif not is_admin:
                raise RuntimeError(f'User {admin_user!r} is not an admin. See --admin-set option')

            purges = dict()
            def wait_and_purge_room(room_id=None, event_id=None):
                """ Wait for available slots in parallel_purges and purge room

                If room_id is None, just wait for current purges to complete and return
                If event_id is None, purge all events in room
                """
                while len(purges) >= (parallel_purges if room_id else 1):
                    # wait and clear completed purges
                    time.sleep(1)
                    for _room_id, purge_id in list(purges.items()):
                        response = session.get(
                            urljoin(
                                server,
                                '/_matrix/client/r0/admin/purge_history_status/' + quote(purge_id),
                            ),
                            params={'access_token': admin_access_token},
                        )
                        assert response.status_code == 200, f'{response!r} => {response.text!r}'
                        if response.json()['status'] != 'active':
                            click.secho(f'Finished purge: room {_room_id!r}, purge {purge_id!r}')
                            purges.pop(_room_id)

                if not room_id:
                    return

                body = {'delete_local_events': True}
                if event_id:
                    body['purge_up_to_event_id'] = event_id
                else:
                    body['purge_up_to_ts'] = int(time.time() * 1000)
                response = session.post(
                    urljoin(server, '/_matrix/client/r0/admin/purge_history/' + quote(room_id)),
                    params={'access_token': admin_access_token},
                    json=body,
                )
                if response.status_code == 200:
                    purge_id = response.json()['purge_id']
                    purges[room_id] = purge_id
                    return purge_id

            if not keep_newer and not keep_min_msgs:
                click.confirm(
                    'No --keep-newer nor --keep-min-msgs option provided. Purge all history?',
                    abort=True,
                )

            ts_ms = None
            if keep_newer:
                ts = datetime.datetime.now() - datetime.timedelta(keep_newer)
                ts_ms = int(ts.timestamp() * 1000)

            cur.execute('SELECT room_id FROM rooms ;')
            all_rooms = {row for row, in cur}

            click.secho(f'Processing {len(all_rooms)} rooms')
            for room_id in all_rooms:
                # no --keep-min-msgs nor --keep-newer, purge everything
                if not keep_newer and not keep_min_msgs:
                    wait_and_purge_room(room_id)
                    continue
                cur.execute(
                    f"""
                    SELECT event_id FROM (
                        SELECT event_id,
                            received_ts,
                            COUNT(*) OVER (ORDER BY received_ts DESC) AS msg_count_above
                        FROM events
                        WHERE room_id=%(room_id)s AND type='m.room.message'
                        ORDER BY received_ts DESC
                    ) t WHERE true
                    {'AND received_ts < %(ts_ms)s' if keep_newer else ''}
                    {'AND msg_count_above > %(keep_min_msgs)s' if keep_min_msgs else ''}
                    LIMIT 1 ;""",
                    {
                        'room_id': room_id,
                        'ts_ms': ts_ms,
                        'keep_min_msgs': keep_min_msgs,
                    },
                )
                if cur.rowcount:
                    event_id, = cur.fetchone()
                    wait_and_purge_room(room_id, event_id)
                # else: room doesn't have messages eligible for purging, skip

            wait_and_purge_room(None)

        if post_sql:
            click.secho(f'Running {post_sql.name!r}')
            with psycopg2.connect(db_uri) as db, db.cursor() as cur:
                cur.execute(post_sql.read())
                click.secho(f'Results {cur.rowcount}:')
                for i, row in enumerate(cur):
                    click.secho(f'{i}: {row}')

    finally:
        if docker_restart_label:
            client = docker.from_env()
            for container in client.containers.list():
                if container.attrs['State']['Status'] != 'running' or\
                        not container.attrs['Config']['Labels'].get(docker_restart_label):
                    continue

                try:
                    # parse container's env vars
                    env_vars = dict(
                        itemgetter(0, 2)(e.partition('='))
                        for e in container.attrs['Config']['Env']
                    )
                    remote_config_file = env_vars.get(
                        'URL_KNOWN_FEDERATION_SERVERS',
                    ) or URL_KNOWN_FEDERATION_SERVERS_DEFAULT

                    # fetch remote file
                    remote_whitelist = yaml.load(requests.get(remote_config_file).text)

                    # fetch local list from container's synapse config
                    local_whitelist = yaml.load(
                        container.exec_run(['cat', SYNAPSE_CONFIG_PATH]).output,
                    )['federation_domain_whitelist']

                    # if list didn't change, don't proceed to restart container
                    if local_whitelist and remote_whitelist == local_whitelist:
                        continue

                    click.secho(
                        f'Whitelist changed. Restarting. new_list={remote_whitelist!r}',
                    )
                except (
                        KeyError,
                        IndexError,
                        requests.RequestException,
                        yaml.scanner.ScannerError,
                ) as ex:
                    click.secho(
                        f'An error ocurred while fetching whitelists: {ex!r}\n'
                        'Restarting anyway',
                        err=True,
                    )
                # restart container
                container.restart(timeout=30)


if __name__ == '__main__':
    purge(auto_envvar_prefix='MATRIX')
